src/AeroSharp/DataAccess/MapAccess/Parsers/MapParserWithSerializer.cs

```diff
@@ -0,0 +1,73 @@
+﻿using AeroSharp.DataAccess.Exceptions;
+using AeroSharp.Serialization;
+using Aerospike.Client;
+using System;
+using System.Collections.Generic;
+
+namespace AeroSharp.DataAccess.MapAccess.Parsers;
+
+/// <summary>
+/// Parses map data from Aerospike records containing serialized values.
+/// </summary>
+internal sealed class MapParserWithSerializer : IMapParser
+{
+    private readonly IMapBinParser _mapBinParser;
+    private readonly ISerializer _valueSerializer;
+
+    public MapParserWithSerializer(IMapBinParser mapBinParser, ISerializer valueSerializer)
+    {
+        _mapBinParser = mapBinParser;
+        _valueSerializer = valueSerializer;
+    }
+
+    public KeyValuePair<TKey, TValue> Parse<TKey, TValue>(Record record, string bin)
+    {
+        var mapEntry = _mapBinParser.ParseOne(record, bin);
+        var key = ExtractKey<TKey>(mapEntry, bin);
+        var byteValue = ExtractValue(mapEntry, bin);
+
+        return DeserializeValue<TKey, TValue>(key, byteValue);
+    }
+
+    private static TKey ExtractKey<TKey>(KeyValuePair<object, object> mapEntry, string bin)
+    {
+        try
+        {
+            return (TKey)mapEntry.Key;
+        }
+        catch (Exception exception)
+        {
+            throw new UnexpectedDataFormatException(
+                $"Map key in bin \"{bin}\" is not a {typeof(TKey).FullName}.",
+                exception
+            );
+        }
+    }
+
+    private static byte[] ExtractValue(KeyValuePair<object, object> mapEntry, string bin)
+    {
+        try
+        {
+            return (byte[])mapEntry.Value;
+        }
+        catch (Exception exception)
+        {
+            throw new UnexpectedDataFormatException(
+                $"Value in bin \"{bin}\" is not a {typeof(byte[]).FullName}.",
+                exception
+            );
+        }
+    }
+
+    private KeyValuePair<TKey, TValue> DeserializeValue<TKey, TValue>(TKey key, byte[] byteValue)
+    {
+        try
+        {
+            return new KeyValuePair<TKey, TValue>(key, _valueSerializer.Deserialize<TValue>(byteValue));
+        }
+        catch (Exception exception)
+        {
+            throw new DeserializationException("Error deserializing bin blob.", exception);
```

Comment:
I realize this method doesn't know what the bin name is, but I think the error might be helpful if it did. This is similar to ExtractValue, I think. What do you think?

Reply:
Updated this to take in bin as an arg and include it in the exception message.